File: cpp/examples/sdk_registration_demo.cpp
#include "horus/robot/robot.hpp"
#include "horus/robot/sensors.hpp"
#include "horus/utils/branding.hpp"
#include "horus/utils/logging.hpp"

#include <any>
#include <map>
#include <memory>
#include <optional>
#include <string>
#include <vector>

namespace {
std::optional<std::string> get_arg(int argc, char** argv, const std::string& name) {
    for (int i = 1; i < argc - 1; ++i) {
        if (argv[i] == name) {
            return std::string(argv[i + 1]);
        }
    }
    return std::nullopt;
}
} // namespace

int main(int argc, char** argv) {
    horus::utils::show_ascii_art();

    const auto robot_count = get_arg(argc, argv, "--robot-count").has_value()
                                 ? std::max(1, std::stoi(*get_arg(argc, argv, "--robot-count")))
                                 : 4;
    const auto with_camera = !get_arg(argc, argv, "--with-camera").has_value() ||
                             (*get_arg(argc, argv, "--with-camera") != "false");
    const auto with_occupancy = !get_arg(argc, argv, "--with-occupancy-grid").has_value() ||
                                (*get_arg(argc, argv, "--with-occupancy-grid") != "false");
    const auto workspace_scale = get_arg(argc, argv, "--workspace-scale")
                                     ? std::optional<double>(std::stod(*get_arg(argc, argv, "--workspace-scale")))
                                     : std::nullopt;

    horus::utils::print_step("Defining Robot Configuration...");

    std::vector<horus::robot::Robot> robots;
    std::vector<horus::dataviz::DataViz> datavizs;
    std::vector<horus::robot::Robot*> robot_ptrs;

    robots.reserve(static_cast<std::size_t>(robot_count));
    datavizs.reserve(static_cast<std::size_t>(robot_count));
    robot_ptrs.reserve(static_cast<std::size_t>(robot_count));

    for (int i = 0; i < robot_count; ++i) {
        const auto name = "test_bot_" + std::to_string(i + 1);
        horus::robot::RobotDimensions dims{
            0.8 + (0.1 * i),
            0.6 + (0.05 * i),
            0.4 + (0.03 * i),
        };
        robots.emplace_back(name, horus::core::RobotType::WHEELED, dims);
        auto& robot = robots.back();
        robot.configure_ros_binding("prefixed", "prefixed", "base_link");
        robot.configure_robot_manager(true, true, true, true);
        robot.configure_workspace_compass(true, 8088, "auto");
        robot.configure_workspace_tutorial("multi_robot_intro");
        if (i == 0) {
            robot.configure_local_body_model("rosbot", true);
        }
        robot.set_metadata(
            "teleop_config",
            std::map<std::string, std::any>{
                {"enabled", true},
                {"robot_profile", std::string("wheeled")},
                {"response_mode", std::string("analog")},
                {"publish_rate_hz", 60.0},
                {"axes",
                 std::map<std::string, std::any>{
                     {"deadzone", 0.12},
                     {"linear_xy_max_mps", 1.1 + (0.05 * i)},
                     {"angular_z_max_rps", 1.4},
                 }},
            });
        robot.set_metadata(
            "task_config",
            std::map<std::string, std::any>{
                {"go_to_point",
                 std::map<std::string, std::any>{
                     {"frame_id", std::string("map")},
                     {"position_tolerance_m", 0.18},
                 }},
                {"waypoint",
                 std::map<std::string, std::any>{
                     {"frame_id", std::string("map")},
                     {"position_tolerance_m", 0.25},
                 }},
            });

        if (with_camera) {
            auto camera = std::make_shared<horus::robot::Camera>(
                "front_camera",
                name + "/camera_link",
                "/" + name + "/camera/image_raw/compressed");
            camera->set_streaming_type("ros");
            camera->set_minimap_streaming_type("ros");
            camera->set_teleop_streaming_type("webrtc");
            camera->set_startup_mode("minimap");
            camera->set_resolution({320, 180});
            camera->set_fps(6);
            camera->set_encoding("jpeg");
            camera->set_minimap_topic("/" + name + "/camera/minimap/compressed");
            camera->set_teleop_topic("/" + name + "/camera/image_raw/compressed");
            camera->set_minimap_image_type("compressed");
            camera->set_teleop_image_type("compressed");
            camera->set_minimap_max_fps(12);
            camera->add_metadata("image_type", std::string("compressed"));
            camera->add_metadata("streaming_type", std::string("ros"));
            camera->add_metadata("minimap_streaming_type", std::string("ros"));
            camera->add_metadata("teleop_streaming_type", std::string("webrtc"));
            camera->add_metadata("startup_mode", std::string("minimap"));
            camera->add_metadata("display_mode", std::string("projected"));
            camera->add_metadata("projection_target_frame", name + "/camera_link");
            camera->add_metadata(
                "projected_scale_multiplier",
                std::vector<float>{1.0f, 1.0f, 1.0f});
            robot.add_sensor(camera);
        }

        auto dataviz = robot.create_dataviz();
        robot.add_path_planning_to_dataviz(
            *dataviz,
            robot.resolve_topic("global_path"),
            robot.resolve_topic("local_path"),
            robot.resolve_topic("trajectory"));
        robot.add_navigation_safety_to_dataviz(*dataviz);
        if (with_occupancy) {
            horus::dataviz::RenderOptions render_options;
            render_options["show_unknown_space"] = true;
            dataviz->add_occupancy_grid("/map", "map", render_options);
        }
        dataviz->add_3d_map(
            "/" + name + "/map_points",
            "map",
            {
                {"point_size", 0.01},
                {"max_points_per_frame", 60000},
                {"enable_adaptive_quality", true},
                {"render_mode", std::string("opaque_fast")},
            });
        dataviz->add_3d_mesh(
            "/" + name + "/map_mesh",
            "map",
            {
                {"max_triangles", 200000},
                {"source_coordinate_space", std::string("enu")},
            });
        dataviz->add_3d_octomap(
            "/" + name + "/octomap_mesh",
            "map",
            {
                {"native_topic", "/" + name + "/octomap_binary"},
                {"native_frame", std::string("map")},
                {"render_mode", std::string("surface_mesh")},
            });
        dataviz->add_semantic_box(
            name + "_dock",
            "Dock",
            {static_cast<float>(i), 0.0f, 0.0f},
            {0.6f, 0.4f, 0.3f});
        datavizs.push_back(*dataviz);
        robot_ptrs.push_back(&robot);
    }

    horus::utils::print_step("Registering robots...");
    auto [ok, result] =
        horus::robot::register_robots(robot_ptrs, datavizs, 10.0, true, true, workspace_scale);
    if (!ok) {
        horus::utils::print_error("Registration failed");
        return 1;
    }

    horus::utils::print_success("Registration completed");
    if (const auto it = result.find("results"); it != result.end()) {
        if (it->second.type() == typeid(std::vector<std::map<std::string, std::any>>)) {
            const auto entries =
                std::any_cast<std::vector<std::map<std::string, std::any>>>(it->second);
            horus::utils::print_info("Registered robots: " + std::to_string(entries.size()));
        }
    }
    return 0;
}

File: rust/examples/sdk_registration_demo.rs
use clap::Parser;
use horus::core::types::RobotType;
use horus::robot::{register_robots, Robot, RobotDimensions};
use horus::sensors::Camera;
use std::collections::HashMap;
use std::sync::Arc;

#[derive(Debug, Parser)]
#[command(about = "Register multiple robots with Horus (Rust parity demo).")]
struct Args {
    #[arg(long, default_value = "test_bot")]
    robot_name: String,
    #[arg(long, default_value = "")]
    robot_names: String,
    #[arg(long, default_value_t = 10)]
    robot_count: usize,
    #[arg(long, default_value_t = true)]
    keep_alive: bool,
    #[arg(long, default_value_t = 0.8)]
    length: f64,
    #[arg(long, default_value_t = 0.6)]
    width: f64,
    #[arg(long, default_value_t = 0.4)]
    height: f64,
    #[arg(long, default_value_t = true)]
    with_camera: bool,
    #[arg(long, default_value_t = true)]
    with_occupancy_grid: bool,
    #[arg(long, default_value = "compressed")]
    camera_image_type: String,
    #[arg(long)]
    camera_streaming_type: Option<String>,
    #[arg(long)]
    camera_minimap_streaming_type: Option<String>,
    #[arg(long)]
    camera_teleop_streaming_type: Option<String>,
    #[arg(long, default_value = "minimap")]
    camera_startup_mode: String,
    #[arg(long, default_value = "160x90,192x108,224x126,256x144,320x180,426x240")]
    camera_resolutions: String,
    #[arg(long, default_value_t = true)]
    vary_camera_resolution: bool,
    #[arg(long, default_value = "/map")]
    occupancy_topic: String,
    #[arg(long, default_value = "map")]
    occupancy_frame: String,
    #[arg(long)]
    occupancy_position_scale: Option<f64>,
    #[arg(long)]
    workspace_scale: Option<f64>,
}

fn parse_resolution_list(raw_value: &str) -> Vec<(u32, u32)> {
    raw_value
        .split(',')
        .filter_map(|token| {
            let item = token.trim().to_ascii_lowercase();
            let (left, right) = item.split_once('x')?;
            let width = left.parse::<u32>().ok()?.max(16);
            let height = right.parse::<u32>().ok()?.max(16);
            Some((width, height))
        })
        .collect()
}

fn resolve_robot_names(args: &Args) -> Vec<String> {
    if !args.robot_names.trim().is_empty() {
        let explicit: Vec<String> = args
            .robot_names
            .split(',')
            .map(str::trim)
            .filter(|name| !name.is_empty())
            .map(ToString::to_string)
            .collect();
        if !explicit.is_empty() {
            return explicit;
        }
    }
    if args.robot_count <= 1 {
        return vec![args.robot_name.clone()];
    }
    (0..args.robot_count)
        .map(|index| format!("{}_{}", args.robot_name, index + 1))
        .collect()
}

fn main() {
    let args = Args::parse();
    let robot_names = resolve_robot_names(&args);
    let mut resolutions = parse_resolution_list(&args.camera_resolutions);
    if resolutions.is_empty() {
        resolutions.push((160, 90));
    }

    let legacy_streaming = args
        .camera_streaming_type
        .clone()
        .unwrap_or_default()
        .trim()
        .to_ascii_lowercase();
    let minimap_streaming = args
        .camera_minimap_streaming_type
        .clone()
        .unwrap_or_else(|| legacy_streaming.clone());
    let minimap_streaming = if minimap_streaming.is_empty() {
        "ros".to_string()
    } else {
        minimap_streaming
    };
    let teleop_streaming = args
        .camera_teleop_streaming_type
        .clone()
        .unwrap_or_else(|| legacy_streaming.clone());
    let teleop_streaming = if teleop_streaming.is_empty() {
        "webrtc".to_string()
    } else {
        teleop_streaming
    };
    let startup_mode = {
        let normalized = args.camera_startup_mode.trim().to_ascii_lowercase();
        if normalized == "teleop" {
            "teleop".to_string()
        } else {
            "minimap".to_string()
        }
    };
    let legacy_payload_streaming = if legacy_streaming.is_empty() {
        minimap_streaming.clone()
    } else {
        legacy_streaming
    };

    let mut robots = Vec::new();
    let mut datavizs = Vec::new();

    for (index, name) in robot_names.iter().enumerate() {
        let dimensions = RobotDimensions::new(
            args.length + (index as f64 * 0.1),
            args.width + (index as f64 * 0.05),
            args.height + (index as f64 * 0.03),
        );
        let mut robot = Robot::with_dimensions(name.clone(), RobotType::Wheeled, dimensions);
        robot.configure_ros_binding("prefixed", "prefixed", Some("base_link"));
        robot.configure_robot_manager(true, true, true, true);
        robot.configure_workspace_compass(true, 8088, "auto");
        robot.configure_workspace_tutorial("multi_robot_intro");
        if index == 0 {
            robot.configure_local_body_model("rosbot", true);
        }
        robot.set_metadata(
            "teleop_config",
            serde_json::json!({
                "enabled": true,
                "robot_profile": "wheeled",
                "response_mode": "analog",
                "publish_rate_hz": 60.0,
                "axes": {
                    "deadzone": 0.12,
                    "linear_xy_max_mps": 1.1 + (index as f64 * 0.05),
                    "angular_z_max_rps": 1.4
                }
            }),
        );
        robot.set_metadata(
            "task_config",
            serde_json::json!({
                "go_to_point": {
                    "frame_id": "map",
                    "position_tolerance_m": 0.18
                },
                "waypoint": {
                    "frame_id": "map",
                    "position_tolerance_m": 0.25
                }
            }),
        );

        if args.with_camera {
            let image_type = args.camera_image_type.trim().to_ascii_lowercase();
            let (topic, encoding) = if image_type == "compressed" {
                (format!("/{name}/camera/image_raw/compressed"), "jpeg")
            } else {
                (format!("/{name}/camera/image_raw"), "rgb8")
            };

            let (width, height) = if args.vary_camera_resolution {
                resolutions[index % resolutions.len()]
            } else {
                resolutions[0]
            };

            let mut camera = Camera::try_new_with_profiles(
                "front_camera",
                format!("{name}/camera_link"),
                topic,
                &legacy_payload_streaming,
                &minimap_streaming,
                &teleop_streaming,
                &startup_mode,
            )
            .unwrap_or_else(|_| {
                Camera::new(
                    "front_camera",
                    format!("{name}/camera_link"),
                    format!("/{name}/camera/image_raw"),
                )
            });

            camera.resolution = (width, height);
            camera.fps = 6;
            camera.encoding = encoding.to_string();
            camera.minimap_topic = format!("/{name}/camera/minimap/compressed");
            camera.teleop_topic = format!("/{name}/camera/image_raw/compressed");
            camera.minimap_image_type = "compressed".to_string();
            camera.teleop_image_type = "compressed".to_string();
            camera.minimap_max_fps = 12;
            camera.add_metadata("image_type", serde_json::json!(image_type));
            camera.add_metadata(
                "streaming_type",
                serde_json::json!(legacy_payload_streaming),
            );
            camera.add_metadata(
                "minimap_streaming_type",
                serde_json::json!(minimap_streaming.clone()),
            );
            camera.add_metadata(
                "teleop_streaming_type",
                serde_json::json!(teleop_streaming.clone()),
            );
            camera.add_metadata("startup_mode", serde_json::json!(startup_mode.clone()));
            camera.add_metadata("display_mode", serde_json::json!("projected"));
            camera.add_metadata("use_tf", serde_json::json!(true));
            camera.add_metadata(
                "projection_target_frame",
                serde_json::json!(format!("{name}/camera_link")),
            );
            camera.add_metadata(
                "projected_scale_multiplier",
                serde_json::json!({"x": 1.0, "y": 1.0, "z": 1.0}),
            );

            if robot.add_sensor(Arc::new(camera)).is_err() {
                eprintln!("failed to add camera sensor for {}", name);
            }
        }

        let mut dataviz = robot.create_dataviz(None);
        if args.with_occupancy_grid {
            let mut render_options = std::collections::HashMap::new();
            if let Some(scale) = args.occupancy_position_scale {
                render_options.insert(
                    "position_scale".to_string(),
                    serde_json::json!(scale.max(0.01)),
                );
            }
            dataviz.add_occupancy_grid(
                &args.occupancy_topic,
                &args.occupancy_frame,
                Some(render_options),
            );
        }
        let global_path = robot.resolve_topic("global_path");
        let local_path = robot.resolve_topic("local_path");
        let trajectory = robot.resolve_topic("trajectory");
        robot.add_path_planning_to_dataviz(
            &mut dataviz,
            Some(&global_path),
            Some(&local_path),
            Some(&trajectory),
        );
        robot.add_navigation_safety_to_dataviz(&mut dataviz);

        let mut point_options = HashMap::new();
        point_options.insert("point_size".to_string(), serde_json::json!(0.01));
        point_options.insert("max_points_per_frame".to_string(), serde_json::json!(60000));
        point_options.insert(
            "enable_adaptive_quality".to_string(),
            serde_json::json!(true),
        );
        point_options.insert("render_mode".to_string(), serde_json::json!("opaque_fast"));
        dataviz.add_3d_map(&format!("/{name}/map_points"), "map", Some(point_options));

        let mut mesh_options = HashMap::new();
        mesh_options.insert("max_triangles".to_string(), serde_json::json!(200000));
        mesh_options.insert(
            "source_coordinate_space".to_string(),
            serde_json::json!("enu"),
        );
        dataviz.add_3d_mesh(&format!("/{name}/map_mesh"), "map", Some(mesh_options));

        let mut octomap_options = HashMap::new();
        octomap_options.insert(
            "native_topic".to_string(),
            serde_json::json!(format!("/{name}/octomap_binary")),
        );
        octomap_options.insert("native_frame".to_string(), serde_json::json!("map"));
        octomap_options.insert("render_mode".to_string(), serde_json::json!("surface_mesh"));
        dataviz.add_3d_octomap(
            &format!("/{name}/octomap_mesh"),
            "map",
            Some(octomap_options),
        );
        dataviz.add_semantic_box(
            &format!("{name}_dock"),
            "Dock",
            (index as f32, 0.0, 0.0),
            (0.6, 0.4, 0.3),
            "map",
            None,
        );

        datavizs.push(dataviz);
        robots.push(robot);
    }

    println!("Registering {} robot(s)...", robots.len());
    let (ok, result) = register_robots(
        &mut robots,
        Some(datavizs),
        args.keep_alive,
        true,
        10.0,
        args.workspace_scale,
    );
    if !ok {
        eprintln!("Registration failed: {result}");
        return;
    }
    println!("Registration result: {result}");
}
